tests/components/snooz/test_fan.py

```diff
@@ -0,0 +1,268 @@
+
+
+async def test_set_0_percentage_turns_off(
+    hass: HomeAssistant, snooz_fan_entity_id: str
+):
+    """Test turning off the device by setting the percentage/volume to 0."""
+    await hass.services.async_call(
+        fan.DOMAIN,
+        fan.SERVICE_TURN_ON,
+        {ATTR_ENTITY_ID: [snooz_fan_entity_id], fan.ATTR_PERCENTAGE: 66},
+        blocking=True,
+    )
+
+    await hass.services.async_call(
+        fan.DOMAIN,
+        fan.SERVICE_SET_PERCENTAGE,
+        {ATTR_ENTITY_ID: [snooz_fan_entity_id], fan.ATTR_PERCENTAGE: 0},
+        blocking=True,
+    )
+
+    state = hass.states.get(snooz_fan_entity_id)
+    assert state.state == STATE_OFF
+    # doesn't overwrite percentage when turning off
+    assert state.attributes[fan.ATTR_PERCENTAGE] == 66
+    assert ATTR_ASSUMED_STATE not in state.attributes
+
+
+async def test_turn_off(hass: HomeAssistant, snooz_fan_entity_id: str):
+    """Test turning off the device."""
+    await hass.services.async_call(
+        fan.DOMAIN,
+        fan.SERVICE_TURN_OFF,
+        {ATTR_ENTITY_ID: [snooz_fan_entity_id]},
+        blocking=True,
+    )
+
+    state = hass.states.get(snooz_fan_entity_id)
+    assert state.state == STATE_OFF
+    assert ATTR_ASSUMED_STATE not in state.attributes
+
+
+async def test_push_events(
+    hass: HomeAssistant, mock_connected_snooz: SnoozFixture, snooz_fan_entity_id: str
+):
+    """Test state update events from snooz device."""
+    mock_connected_snooz.device.trigger_state(SnoozDeviceState(False, 64))
+
+    state = hass.states.get(snooz_fan_entity_id)
+    assert ATTR_ASSUMED_STATE not in state.attributes
+    assert state.state == STATE_OFF
+    assert state.attributes[fan.ATTR_PERCENTAGE] == 64
+
+    mock_connected_snooz.device.trigger_state(SnoozDeviceState(True, 12))
+
+    state = hass.states.get(snooz_fan_entity_id)
+    assert ATTR_ASSUMED_STATE not in state.attributes
+    assert state.state == STATE_ON
+    assert state.attributes[fan.ATTR_PERCENTAGE] == 12
+
+    mock_connected_snooz.device.trigger_disconnect()
+
+    state = hass.states.get(snooz_fan_entity_id)
+    assert state.attributes[ATTR_ASSUMED_STATE] is True
+
+
+@pytest.mark.parametrize(
+    "fan_state,percentage",
+    [
+        (STATE_OFF, 1),
+        (STATE_OFF, 22),
+        (STATE_OFF, 50),
+        (STATE_OFF, 99),
+        (STATE_OFF, 100),
+        (STATE_ON, 1),
+        (STATE_ON, 22),
+        (STATE_ON, 50),
+        (STATE_ON, 99),
+        (STATE_ON, 100),
+    ],
+)
+async def test_restore_state(hass: HomeAssistant, fan_state: str, percentage: int):
+    """Tests restoring entity state."""
+    disconnected_device = await create_mock_snooz(
+        connected=False, initial_state=UnknownSnoozState
+    )
+    restored_state = State(
+        "fan.test_restored", fan_state, {fan.ATTR_PERCENTAGE: percentage}
+    )
+
+    with patch(
+        "homeassistant.helpers.restore_state.RestoreEntity.async_get_last_state",
+        return_value=restored_state,
+    ):
```

Comment:
We aren't allowed to patch the underlying entity class.

Instead setup the integration, tear it down, and start it back up to make sure its restored

https://developers.home-assistant.io/docs/development_testing?_highlight=testing#writing-tests-for-integrations